driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -581,9 +583,9 @@ boolean setPool(HostConnectionPool pool) {
      * The connection should generally not be reused after that.
      */
     void release() {
-        HostConnectionPool pool = poolRef.get();
-        if (pool != null)
-            pool.returnConnection(this);
+        Owner owner = ownerRef.get();
+        if (owner instanceof HostConnectionPool)
```

Comment:
It sounds like you tried this scenario with the harness in `HostConnectionPoolTest`. The problem is that this harness is not well-suited to test for errors: in real life, if a connection gets closed, the way inFlight gets decreased is by failing all pending `ResponseHandlers`, which will themselves release the connection (see `SpeculativeExecution#onException`).
In `HostConnectionPoolTest` there are no handlers, we're just borrowing and releasing connections manually. In fact I'm not even sure why totalInFlight was decremented.
One thing we could do in the test is use a simple `ResponseHandler` implementation instead of borrowing and releasing manually.